Add ability to copy assistant code block to clipboard or insert into editor, without manual selection

thataboy opened the pull request:
Some notes:

- You can put the cursor on the start or end line with triple backticks, it doesn't actually have to be inside the block.
- Placing the cursor outside of a code block does nothing.
- Code blocks are determined by counting triple backticks pairs from either start or end of buffer, and nothing else.
- If you manually select something, the selection takes precedence over any code blocks.

Release Notes:

- Added the ability to copy surrounding code blocks in the assistant panel into the clipboard, or inserting them directly into the editor, without manually selecting. Place cursor anywhere in a code block (marked by triple backticks) and use the `assistant::CopyCode` action (`cmd-k c` / `ctrl-k c`) to copy to the clipboard, or the `assistant::InsertIntoEditor` action (`cmd-<` / `ctrl-<`) to insert into editor.

thataboy committed:
add ability to copy assistant code block to clipboard or insert into editor, without manual selection

thataboy committed:
add key bindings for assistant::CopySelection

mrnugget commented on crates/assistant/src/assistant_panel.rs:
```diff
@@ -3121,6 +3198,36 @@ impl ContextEditor {
+
+        if !text.is_empty() {
+            cx.write_to_clipboard(ClipboardItem::new_string(text));
+            if is_code_block {
```
Nitpick: why do we only show this when it's a code block?

thataboy replied:
> Nitpick: why do we only show this when it's a code block?

I figured because copying while you have a selection is quite normal and you don't expect to get any feedback for it, whereas copying "nothing" is a little non standard. But yeah, maybe the distinction isn't necessary

mrnugget replied:
Yeah, I'm a bit unsure about the mechanics. I guess my question is: why do we need a new command if there's a selection? If there's a selection, you can `cmd-c` to copy, right? Should we maybe make your new command a no-op if there is a selection? Or should it expand the selection and then copy?

thataboy replied:
I didn't do a no-op because as I said I want to be consistent with the new insert into editor, which is: insert selection if there is one (old behavior), else insert the code block if inside a code block (new behavior).

Visually selecting the block is a good idea and it's what I tried to do but I couldn't figure out the code. Specifically, I don't know how to get the `cx: &mut ViewContext<Self>` param for `Editor > change_selections` when I have `cx: &mut ViewContext<Workspace>`. Can you help? If I can select the code block, that will make it clear what's happening and we can get rid of the toast message.

mrnugget replied:
> I didn't do a no-op because as I said I want to be consistent with the new insert into editor, which is: insert selection if there is one (old behavior), else insert the code block if inside a code block (new behavior).

Ah, I see. I didn't realise that you also modified the existing action. Sorry!



> Specifically, I don't know how to get the `cx: &mut ViewContext<Self>` param for `Editor > change_selections` when I have `cx: &mut ViewContext<Workspace>`. Can you help? If I can select the code block, that will make it clear what's happening and we can get rid of the toast message.

Sure! The `context_editor_view.read(cx).editor` is a `View<Editor>`. If you call `update` on that, you can pass in a callback that receives `editor` and `cx`, which is the `ViewContext<Editor>`. Like this:

```rust
fn get_selection_or_code_block(
    context_editor_view: &View<ContextEditor>,
    cx: &mut ViewContext<Workspace>,
) -> (String, bool) {
    // (you don't need to specify the types, I'm just doing it here for illustration)
    context_editor_view.read(cx).editor.update(cx, |editor: &mut Editor, cx: &mut ViewContext<Editor>| {
        
    });
    // ...
}
```

thataboy replied:
thanks. I needed to throw in a clone() after read(cx).editor and rearrange the code a bit to make the borrow checker happy, but it worked. Will push an update later today :)

thataboy committed:
Select code block when inserting for clearer visual feedback

thataboy committed:
Rename `assistant::CopySelection` to `assistant::CopyCode`

mrnugget committed:
Use tree-sitter syntax tree to find code blocks

This changes the approach to use the tree-sitter syntax tree to find
code blocks instead of parsing the buffer by hand.

This also adds some tests.

Co-authored-by: Bennet <bennet@zed.dev>

mrnugget committed:
Fix typo

Co-authored-by: Bennet <bennet@zed.dev>

mrnugget committed:
Fix compile error after rebase

mrnugget commented:
Hey! @bennetbo and I hopped on this branch and after trying to refactor the code a little bit (there were also some bugs, where it wouldn't correctly find code blocks if there were multiple of them), we decided that the easier way would be to make use of the already created syntax tree which we have from tree-sitter. So the new approach should be a more efficient.

Thank you so much for this!

mrnugget committed:
Simplify code and fix an off-by-one bug

Co-authored-by: Bennet <bennet@zed.dev>

kzhui125 commented:
@thataboy  Is `cmd-<` already a shortcut for moving the cursor to the start of the line?

thataboy commented:
> @thataboy Is `cmd-<` already a shortcut for moving the cursor to the start of the line?

It was already the keybinding for inserting into the editor. I did not add it.
